### ISSUE TITLE
Implement Student Attendance Report

### ISSUE THREAD

**Issue by ShanikaEdiriweera** [tharakanandakumara/sign#6]

**Comment by jakkub21** [comment]
@ShanikaEdiriweera hey,
Any adictional information for this issue? I clone project but i'm not able to log in throught your login system so can't see what's going on at the admin panel. 

But i would love to contribute on this issue to help out. Is there any information about your backend API or how can i pull data from the server to populate report and what data do you want to put in this report?

Thanks for answer.

**Comment by ShanikaEdiriweera** [comment]
@jakkub21 sure. I will guide you.

This is the backend repo - https://github.com/ShanikaEdiriweera/student-attendance-system-backend Try to set it up

**Comment by jakkub21** [comment]
when i try to run npm start (this error occurred):
[app.log](https://github.com/tharakanandakumara/sign/files/3694691/app.log)

i was checking source code and i was looking for some endpoints for attendance-report or something like that, i found that actually. And it require Date_from and Date_to (not sure about date formats). And probably some admin access to get datas from server.

Maybe you can prepare some test access for me so i can use that API running on your server with current data from my local enviroment. Or if its not allowed to access those data for contributors then we need to run local database using that backend API. But its not working for me. Not sure why, i send you that app.log to check that out.

**Comment by ShanikaEdiriweera** [comment]
Sorry I couldn't mention. Please check the comments here https://github.com/ShanikaEdiriweera/student-attendance-system-backend/issues/1
Have to add config files and run mongodb locally!

**Comment by ShanikaEdiriweera** [comment]
Then you can create an admin user for yourself using the /user API endpoint. check the sample request in samples

**Comment by jakkub21** [comment]
@ShanikaEdiriweera hey, i was busy lately but i will work on this (give me few more days and i will start (still need to run backend API localy)... so i will let you know, about current state with API), and you can assign me this task and i will do my best to complete it.

I also want to suggest to take a look at you current repository, there is a lot of empty files and structure is random or i'm not sure which folder (frontend/ or current root folders are primary). 

Thanks.

**Comment by ShanikaEdiriweera** [comment]
Yeah restructuring the repo would be good thing. I'll add an issue.

Please work in the frontend folder.

**Comment by ShanikaEdiriweera** [comment]
@jakkub21 After setting up backend create an admin user as following.

Here is the sample request to register a user. You can use Postman.
Use the master key you added when configuring.

// POST http://localhost:3000/public/register
// HEADER Authorization: MASTER <MASTER_KEY>
{
	"name": "John Doe",
	"username": "john",
	"password": "11111",
	"email": "john@gmail.com",
	"contact": "777111222",
	"userType": "admin" // "userType": "user"
}

**Comment by ShanikaEdiriweera** [comment]
@jakkub21 are you working on this?

**Comment by jakkub21** [comment]
@ShanikaEdiriweera 
Hey mate,

Backend is running as Intented.

Now i need more samples for how i can make requests, is there any folder that i missed and i can check that??

But, i was expecting then on backend installation database will be created, but table for users for example, was created when i register first user, which is good, but i can't see how can i access some data for our Student Attendance Report Issue

And there are Dialy Report and Monthly Report are you looking for new report called "Student Attendance Report" or you were looking for completing this ones?

But to be honest i was waiting for PR from [#21](https://github.com/tharakanandakumara/sign/issues/21), because he will change structure and i don't want to mess up with that.

But true, i said that i will update you when i set up backend API and i didn't sorry about that.

And yes, i will definitelly work on this, meanwhile, when he is working on rebuilding project structure, we can talk about how report will looks like, and what you are expecting. I guess you want to use Datatables from bootstrap to manage some action with it (sorting, searching, pagination, exporting [PDF/Excel]

Right?

**Comment by ShanikaEdiriweera** [comment]
Yup great to hear that!

What we are looking here is a new report. Lets say you select the grade, class and then student a report will be shown for the particular student.

Here are the endpoints we currently have. 
If we need a new one for this lets create an issue in backend repo!

```
// POST http://localhost:3000/public/login
{
    "username": "user",
    "password": "password"
}
```

**Student Endpoints**
```
GET http://localhost:3000/protected/students
// HEADER Authorization: BEARER <JWT_TOKEN>
```

`// GET http://localhost:3000/public/students/<index>
`
```
// POST http://localhost:3000/public/students
{
    "indexNo": "180000A",     
    "nameWithInitials": "I.U.Ruhunage",   
    "fullName": "Isuru Udara Ruhunage",  
    "DOB": "1992-10-01",
    "gender": "male",
    "address": "Address",
    "grade": "13",
    "section": "13A", //class
    "medium": "Sinhala",
    "contactNo1": "0112345678"
}
```

**Attendance Endpoints**
```
// GET http://localhost:3000/protected/attendance
// To get a date range or single date
// GET http://localhost:3000/protected/attendance?from=2019-05-12&to=2019-05-12
```

```
// POST http://localhost:3000/protected/students/attendance
// POST http://localhost:3000/protected/students/attendance?from=2019-05-12&to=2019-05-12
``` 

**Get attendance for a specific student**
```
// GET http://localhost:3000/protected/students/180009A/attendance
// GET http://localhost:3000/protected/students/180009A/attendance?from=2019-05-12&to=2019-05-12 
```

```
// POST http://localhost:3000/protected/attendance
{
    "indexNo": "180009A",
    "date": "2019-05-11",
    "time": "05:40",
    "timestamp": "2019-05-11T05:40:38.439Z"
}
```

**Report Endpoints**
```
// GET http://localhost:3000/protected/attendance/report
Response:
{
    "report": {
        "2019-05-14": 2,
        "2019-05-13": 2,
        "2019-05-12": 2
    }
}
```

```
// GET http://localhost:3000/protected/attendance/report?from=2019-05-12&to=2019-05-13 
Response:
{
    "reportFrom": "2019-05-12",
    "reportTo": "2019-05-13",
    "noOfDays": 2,
    "report": {
        "2019-05-13": {
            "total": 3,
            "attendanceByGrade": {
                "2": 1,
                "13": 2
            },
            "attendanceByClass": {
                "22": 1,
                "Maths": 2
            }
        },
        "2019-05-12": {
            "total": 5,
            "attendanceByGrade": {
                "1": 1,
                "2": 1,
                "13": 3
            },
            "attendanceByClass": {
                "22": 1,
                "Maths": 3,
                "1A": 1
            }
        }
    }
}

```

**Comment by jakkub21** [comment]
@ShanikaEdiriweera cool.

One question about students??

Can be more then one grade assigned to student in your system?? I mean you were talking about selecting grade, then class then student... but what we can do instead, we can use Bootstrap Select Component to search in list of students (lets say by their full name)

So basically i will use GET /students  endpoint to populate my bootstrap select component with their names: 

```
<searchbox element>...
...
<option value="studentIndex">studentName</option>
...
```

- ofc this is just template for imagination :) And the next element that form can contain are dates (from - to) so you can generate report about student for specific range

What do you think about that? We will skip like 2-3steps in Form (selecting Grade, selecting Class, and then Student/dates... instead you will search for student and ten select dates and final step will be to generate report)

**Comment by ShanikaEdiriweera** [comment]
Grade is like year 1, 2, 3 ....
Class is 1A, 1B, 1C something like that.

So i think we need to select grade class first to narrow down the list of
students

--
Shanika Ediriweera
Department of Computer Science and Engineering
University of Moratuwa

**Comment by ShanikaEdiriweera** [comment]
@jakkub21 hey how is the progress on this?

**Comment by jakkub21** [comment]
Hey @ShanikaEdiriweera , 

actually i was trying to prepare stable version for structure reBuild in "FrontEnd" folder, because there were some part that they are constantly repeating like headers of html page and footer, and some files could be in different folders for example javascripts and css files... But I messed up a little so i will probably roll back to current version and just continue working on that report, i will tell you more about that at the end of this week.

**Comment by jakkub21** [comment]
@ShanikaEdiriweera ok, so i gave up on structure ReBuild, just because i messed up with that. So i just focus on my task :)

We talked about that we will select grade and then class to narrow down student list, but i decided to come up with some improvements.

First of all i checked that there was prepared some frontend design for student profile and i was thinking that we can use that for more then just Attandance Report,

So under User Manager section for Students i add one more item in menu list it's called "Student List"
On this page i list every record from Student endpoint and fill table with some usefull information like fullname, gender, grade and Class and also i added one action for each student and its for redirect to student profile.

Table offering basic Bootstrap Datatable options such as Pagination, Ordering using table heading cells, Filtering using Search Box (filter based on table data), showing entries and also i add option to generate excel document with records in current table view, so basically you can filter data base on student name or even student class and you can export whole list in specific class. (Filtration is done on client side, so we are not fetching data constantly from the backend API)

There wasn't easier way to generate student list based on Class for example or Grade... i would still need to filter data from all of records in Student database table, but i think this solution brings more value to your project and more possibilities.

ok so thats enough informations about student list and table options, i didn't forgot about our issue, so when you finally find your Student, you can use action called "Show Student Profile" which will redirect your admin user to another page where you can see all of his information (student data) + attandance records, with option to export whole table in Excel or even CSV.

I will provide demo version tomorrow so you can see, how it looks and probably even PR.

**Comment by jakkub21** [comment]
Hey @ShanikaEdiriweera,
I send PR #40.

**Comment by ShanikaEdiriweera** [comment]
@jakkub21 This s a good idea.
